torch/csrc/jit/mobile/versioned_operators.cpp

```diff
@@ -0,0 +1,169 @@
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/mobile/versioned_operators.h>
+
+#include <unordered_map>
+
+// This file is to handle operator versioning in mobile.
+// How does it work?
+// At top level, with a tuple of (operator_name, op_version_in_model,
+// bytecode_version_in_model), check the compatibility for that operator. If
+// it's not compatible, throw an error with reason. If it's compatible, do
+// compatibility adaptation if necessary and return the corresponding operator
+// functor that is compatible to this runtime. The logic is implemented in
+// function operator_resolver().
+//
+// Under the hood, a static table,
+// op_version_table, is saved for current runtime. The keys are the operator
+// names and values are a container with version numbers and their metadata. The
+// incoming operator with version is searched in the table to find the match.
+// Note that the size of op_version_table is small because it only contains
+// operators with compatibility updates. For most of the operators, they
+// fall through to the existing implementation with default version 0.
+
+namespace torch {
+namespace jit {
+namespace mobile {
+namespace {
+
+// The version info is limited to combinations of two situations:
+//   * Route to another op name
+//   * Push additional default values
+// It’s fine to update the content of version info, to handle more situations in
+// future. Since it’s implementation details, it should not affect compatibility.
+struct VersionInfo {
```

Comment:
Not sure VersionInfo conveys the meaning.

Maybe VersionResolution, VersionCompatibility, ...?